src/coreclr/src/jit/morph.cpp

```diff
@@ -7090,62 +7095,59 @@ GenTree* Compiler::fgMorphPotentialTailCall(GenTreeCall* call)
 
     const char* failReason      = nullptr;
     bool        canFastTailCall = fgCanFastTailCall(call, &failReason);
-    void*       pfnCopyArgs     = nullptr;
-    // Some tailcalls can (or will) only be done as fast tailcalls.
+
+    CORINFO_TAILCALL_HELPERS tailCallHelpers;
+    bool                     tailCallViaJitHelper = false;
     if (!canFastTailCall)
     {
         if (call->IsImplicitTailCall())
         {
+            // Implicit or opportunistic tail calls are always dispatched via fast tail call
+            // mechanism and never via tail call helper for perf.
             failTailCall(failReason);
             return nullptr;
         }
 
-        // Call is tail. prefixed and cannot be dispatched as a fast tail call.
+        assert(call->IsTailPrefixedCall());
+        assert(call->tailCallInfo != nullptr);
 
+        // We do not currently handle non-standard args except for VSD stubs.
         if (!call->IsVirtualStub() && call->HasNonStandardAddedArgs(this))
         {
-            // Methods with non-standard args will have indirection cell or cookie param passed
-            // in callee trash register (e.g. R11). Tail call helper doesn't preserve it before
-            // tail calling the target method and hence ".tail" prefix on such calls needs to be
-            // ignored.
-            //
-            // Exception to the above rule: although Virtual Stub Dispatch (VSD) calls require
-            // extra stub param (e.g. in R11 on Amd64), they can still be called via tail call helper.
-            // This is done by by adding stubAddr as an additional arg before the original list of
-            // args. For more details see fgMorphTailCallViaHelper() and CreateTailCallCopyArgsThunk()
-            // in Stublinkerx86.cpp.
             failTailCall(
                 "Method with non-standard args passed in callee trash register cannot be tail called via helper");
             return nullptr;
         }
 
-#if defined(_TARGET_ARM64_) || defined(_TARGET_UNIX_)
-        // NYI - TAILCALL_RECURSIVE/TAILCALL_HELPER.
-        // So, bail out if we can't make fast tail call.
-        failTailCall(failReason);
-        return nullptr;
-#elif !defined(_TARGET_X86_)
-        // Ok, now we are _almost_ there. Since this needs helper make sure we
-        // can get the required copy thunk.
-        CorInfoHelperTailCallSpecialHandling handling = CORINFO_TAILCALL_NORMAL;
-        if (call->IsVirtualStub())
+        // On x86 we have a faster mechanism than the general one which we use
+        // in almost all cases.
```

Comment:
If you could expand on the cases that this is not true I think it is worth doing so here.

Reply:
I think such a comment will easily bitrot with future changes (eg. varargs support). The decision is made by`fgCanTailCallViaJitHelper` right after the comment, so it should be simple enough to read this function? Maybe I should make it clearer that this function is making the decision?

Reply:
I think I am good with that, you can also make a note saying see `fgCanTailCallViaJitHelper` for more information.